src/basic/unit-file.c

```diff
@@ -259,27 +259,45 @@ static int directory_name_is_valid(const char *name) {
         return false;
 }
 
-static int unit_file_resolve_symlink(
+int unit_file_resolve_symlink(
                 const char *root_dir,
                 char **search_path,
                 const char *dir,
                 int dirfd,
                 const char *filename,
+                bool resolve_destination_target,
                 char **ret_destination) {
 
-        _cleanup_free_ char *target = NULL, *simplified = NULL, *dst = NULL;
+        _cleanup_free_ char *target = NULL, *simplified = NULL, *dst = NULL, *_dir = NULL;
         int r;
 
-        assert(dir);
-        assert(dirfd >= 0);
+        /* This can be called with either dir+dirfd valid and filename just a name,
+         * or !dir && dirfd==AT_FDCWD, and filename being a full path.
+         *
+         * If resolve_destination_target is true, an absolute path will be returned.
+         * If not, an absolute path is returned for linked unit files, and a relative
+         * path otherwise. */
+
         assert(filename);
         assert(ret_destination);
+        assert(dir || path_is_absolute(filename));
+        assert(dirfd >= 0 || dirfd == AT_FDCWD);
 
         r = readlinkat_malloc(dirfd, filename, &target);
         if (r < 0)
                 return log_warning_errno(r, "Failed to read symlink %s%s%s: %m",
                                          dir, dir ? "/" : "", filename);
 
+        if (!dir) {
+                assert(is_path(filename));
```

Comment:
`path_is_absolute()`?? Moreover, this is not necessary, as it is checked at the beginning of this function.